content/knowledge-base/guides/write-a-composition-function-in-python.md

```diff
@@ -0,0 +1,749 @@
+            )
+
+
+if __name__ == "__main__":
+    unittest.main()
+```
+{{</expand>}}
+
+Run the unit tests using `hatch run`:
+
+```shell {copy-lines="1"}
+hatch run test:unit
+.
+----------------------------------------------------------------------
+Ran 1 test in 0.003s
+
+OK
+```
+
+{{<hint "tip">}}
+[Hatch](https://hatch.pypa.io/) is a Python build tool. It builds Python
+artifacts like wheels. It also manages virtual environments, similar
+to `virtualenv` or `venv`. The `hatch run` command creates a virtual environment
+and runs a command in that environment.
+
+You configure Hatch using `pyproject.toml`.
+{{</hint>}}
+
+You can preview the output of a Composition that uses this function using
+the Crossplane CLI. You don't need a Crossplane control plane to do this.
+
+Create a directory under `function-xbuckets` named `example`, and add the
+three files `xr.yaml`, `composition.yaml`, and `functions.yaml`.
+
+{{<expand "The xr.yaml, composition.yaml and function.yaml files">}}
+
+You can recreate the output below using by running `crossplane beta render` with
+these files.
+
+The `xr.yaml` file contains the composite resource to render:
+
+```yaml
+apiVersion: example.crossplane.io/v1
+kind: XBuckets
+metadata:
+  name: example-buckets
+spec:
+  region: us-east-2
+  names:
+  - crossplane-functions-example-a
+  - crossplane-functions-example-b
+  - crossplane-functions-example-c
+```
+
+The `composition.yaml` file contains the Composition to use to render the
+composite resource:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: create-buckets
+spec:
+  compositeTypeRef:
+    apiVersion: example.crossplane.io/v1
+    kind: XBuckets
+  mode: Pipeline
+  pipeline:
+  - step: create-buckets
+    functionRef:
+      name: function-xbuckets
+```
+
+The `functions.yaml` file contains the Functions the Composition references in
+its pipeline steps:
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: function-xbuckets
+  annotations:
+    render.crossplane.io/runtime: Development
+spec:
+  # The CLI ignores this package when using the Development runtime.
+  # You can set it to any value.
+  package: xpkg.upbound.io/negz/function-xbuckets:v0.1.0
+```
+{{</expand>}}
+
+Note that the Function in `functions.yaml` uses the
+{{<hover label="development" line="6">}}Development{{</hover>}}
+runtime. This tells `crossplane beta render` that your function is running
+locally. It connects to your locally running function instead of using Docker to
+pull and run the function.
+
+```yaml {label="development"}
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: function-xbuckets
+  annotations:
+    render.crossplane.io/runtime: Development
+```
+
+Use `hatch run development` to run your function locally. This tells the
+function to run without encryption or authentication. You should only use it
+during testing and development.
```

Comment:
```suggestion
Use {{<hover line="1" label="run">}}hatch run development{{</hover>}} 
to run your function locally.  
The `development` argument runs the
function without encryption or authentication. 

{{<hint "warning" >}}
Only use `development` during testing and development.
{{< /hint >}}
```

Use `hover` to call draw attention to the command because of the number of nearby code boxes.

Pulls the security warning into a dedicated `warning` box.